File: LeetCode-Java/N-Queens.java
public class Solution {
    public List<String[]> solveNQueens(int n) {
        char[][] record = new char[n][n];
        for(int i = 0;i < n;i ++)
        {
            for(int j = 0;j < n;j ++)
            {
                record[i][j] = '.';
            }
        }
        List<String[]> result = new ArrayList<String[]>();
        solveNQueensPart(record,result,0,n);
        return result;
    }
    
    void solveNQueensPart(char[][] record,List<String[]> result, int row, int n)
    {
        for(int col = 0; col < n;col ++)
        {
            if(isLegal(record,row,col))
            {
                record[row][col] = 'Q';
                if(row == n - 1)
                {
                    String[] tempResult = new String[n];
                    for(int i = 0;i < n;i ++)
                    {
                        tempResult[i] = String.valueOf(record[i]);
                    }
                    result.add(tempResult);
                }
                else
                {
                    solveNQueensPart(record,result,row + 1,n);
                }
                record[row][col] = '.';
            }
        }
        return;
    }
    
    boolean isLegal(char[][] record,int row,int column)
    {
        int n = record.length;
        for(int i = 0;i < n;i ++)
        {
            if(record[row][i] == 'Q' || record[i][column] == 'Q')
                return false;
        }
        int i = 1;
        while(row - i >= 0 && column - i >= 0)
        {
            if(record[row - i][column - i] == 'Q')
                return false;
            i ++;
        }
        i = 1;
        while(row + i < n && column + i < n)
        {
            if(record[row + i][column + i] == 'Q')
                return false;
            i ++;
        }
        i = 1;
        while(row - i >= 0 && column + i < n)
        {
            if(record[row - i][column + i] == 'Q')
                return false;
            i ++;
        }
        i = 1;
        while(row + i < n && column - i >= 0)
        {
            if(record[row + i][column - i] == 'Q')
                return false;
            i ++;
        }
        return true;
    }
}

File: Leetcode-cpp/n_queens.cpp
class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<bool> column(n, false);
        vector<bool> diag1(2 * n + 1, false);
        vector<bool> diag2(2 * n + 1, false);
        vector<vector<string>> result;
        vector<int> record;
        search(result, record, column, diag1, diag2, 0);
        return result;
    }
    
    void search(vector<vector<string>>& result, vector<int>& record, vector<bool>& column,
                vector<bool>& diag1, vector<bool>& diag2, int rowIndex) {
        int n = column.size();
        if (rowIndex ==n) {
            vector<string> tempResult;
            for (int i = 0; i < n; ++i) {
                string rowString = "";
                for (int j = 0; j < n; ++j) {
                    if (record[i] == j) rowString += "Q";
                    else rowString += ".";
                }
                tempResult.push_back(rowString);
            }
            result.push_back(tempResult);
        } else {
            for (int i = 0; i < n; ++i) {
                if (!column[i] && !diag1[rowIndex + i] && !diag2[rowIndex + n - 1 - i]) {
                    column[i] = true;
                    diag1[rowIndex + i] = true;
                    diag2[rowIndex + n - 1 - i] = true;
                    record.push_back(i);
                    search(result, record, column, diag1, diag2, rowIndex + 1);
                    column[i] = false;
                    diag1[rowIndex + i] = false;
                    diag2[rowIndex + n - 1 - i] = false;
                    record.erase(record.end() - 1);
                }
            }
        } 
    }
};